Allow storing outputs in database

janpisl opened the pull request:
Implement functionality for storing output data from run processes in a PostgreSQL or SQLite database

# Overview

This extension allows output data from run processes to be stored in a PostgreSQL or SQLite database. This extension can be enabled in the configuration (section "server", option "store_type"; when set to "db", the extension is active). Credentials for connecting to the database are found in the (new) section "db".

# Contribution Agreement

(as per https://github.com/geopython/pywps/blob/master/CONTRIBUTING.rst#contributions-and-licensing)

I'd like to contribute extension Database Output Storage to PyWPS. I confirm that my contributions to PyWPS will be compatible with the PyWPS license guidelines at the time of contribution.

Commit:
Allow storing outputs in database

Implement functionality for storing output data from run processes in a PostgreSQL or SQLite database

Commit:
data_type is the last parameter

Commit:
fix indent.

Commit:
use lowercase

Commit:
remove redundant import

Commit:
formatting - fix indent, remove whitespaces

Commit:
improved formatting

Commit:
fix import

cehbrecht commented:
@janpisl Please check the travis build. There is an import error:
```
 File "/home/travis/build/geopython/pywps/pywps/inout/storage/db/sqlite.py", line 11, in <module>
    from . import DbStorage
ImportError: cannot import name 'DbStorage'
```

... and pep8 is failing.

Commit:
split code to individual classes

Commit:
improve formatting

janpisl commented:
Hi, MacPingu, thank you for pointing this out. I have edited my code so it
passes the test.

On Mon, Aug 13, 2018 at 3:31 PM, MacPingu <notifications@github.com> wrote:

> @janpisl <https://github.com/janpisl> Please check the travis build.
> There is an import error:
>
>  File "/home/travis/build/geopython/pywps/pywps/inout/storage/db/sqlite.py", line 11, in <module>
>     from . import DbStorage
> ImportError: cannot import name 'DbStorage'
>
> ... and pep8 is failing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geopython/pywps/pull/371#issuecomment-412518819>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AexTMxKt0LP24khknXZezia_5kmk6ziJks5uQX_AgaJpZM4V6DwB>
> .
>

cehbrecht commented:
@janpisl I have tested this PR with my Emu processes and I run into an exception (test with Python 3.6):

```
Traceback (most recent call last):
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/response/execute.py", line 72, in update_status_file                                                       
    self.doc = self._construct_doc()
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/response/execute.py", line 189, in _construct_doc                                                          
    doc = template.render(**self.json)
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/response/execute.py", line 145, in json                                                                    
    data["process"] = self.process.json
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/app/Process.py", line 91, in json                                                                          
    'outputs': [o.json for o in self.outputs],
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/app/Process.py", line 91, in <listcomp>                                                                    
    'outputs': [o.json for o in self.outputs],
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/inout/outputs.py", line 92, in json                                                                        
    data = self._json_reference(data)
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/inout/outputs.py", line 104, in _json_reference                                                            
    data["href"] = self.get_url()
  File "/Users/pingu/Documents/GitHub/birdhouse/pywps/pywps/inout/basic.py", line 711, in get_url                                                                      
    (outtype, storage, url) = self.storage.store(self)
AttributeError: 'NoneType' object has no attribute 'store'
```

I tried this with the Emu [chomsky process](https://github.com/bird-house/emu/blob/master/emu/processes/wps_chomsky.py). But this exception happens with every process having a complex output.

I have used as config: `store_type = file`

Currently you have set `store_type = db` as default. I would keep the file-storage as default.

janpisl commented:
@cehbrecht thank you. i will look into it.

cehbrecht commented:
@janpisl Please have also a look on the codacy check (see link above next to travis). The complains on the tests can be skipped.

Commit:
Use Enum

Use Enum for DATA_TYPE (if not used and DATA_TYPE is only an object, it throws a TypeError: type "object" is not iterable

Commit:
Fix a typo

Commit:
Copy vector data to a schema specified in cfg

Copy vector data to a schema specified in cfg instead of public schema.

Commit:
FIx a typo

Commit:
Remvoe whitespaces

Commit:
Edit comment formatting

Commit:
edit comment  formatting

Commit:
Remove redundant imports

Commit:
Remove obsolete code

Commit:
Remove obsolete import and whitespaces

Commit:
Remove whitespaces

fmigneault commented:
Closing since very outdated. Feel free to resubmit with updates if still relevant.
